packages/graphback-datasync/src/providers/DeltaProvider.ts

```diff
@@ -0,0 +1,164 @@
+import { applyIndexes } from "@graphback/runtime-mongo";
+import { GraphQLObjectType } from 'graphql';
+import { Db, ObjectId, FilterQuery } from "mongodb"
+import { getTableName, GraphbackOperationType } from '@graphback/core';
+
+function getDeltaTableName(baseType: GraphQLObjectType) {
+  return `_${getTableName(baseType)}_delta`
+}
+
+type DeltaOps = GraphbackOperationType.CREATE | GraphbackOperationType.UPDATE | GraphbackOperationType.DELETE;
+
+interface DeltaDB {
+  insertDiff(data: any, op: DeltaOps): Promise<any>
+}
+
+type Hook = (doc: any) => any;
+
+interface HookMap {
+  [GraphbackOperationType.CREATE]: Hook[]
+  [GraphbackOperationType.UPDATE]: Hook[]
+}
+
+/**
+ * Provides an interface for interacting with deltas and stuff
+ */
+export class DeltaProvider implements DeltaDB {
+  protected collectionName: string;
+  protected db: Db;
+  protected hooks:HookMap;
+
+  public constructor(baseType: GraphQLObjectType, db: Db) {
+    this.db = db;
+
+    this.collectionName = getDeltaTableName(baseType);
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    applyIndexes([
+      {
+        key: {
+          "docId": 1,
+          "doc.updatedAt": 1
+        }
+      },
+      {
+        key: {
+          "doc.updatedAt": 1
+        }
+      }
+    ], this.db.collection(this.collectionName));
+  }
+
+
+  public async insertDiff(data: any, op: DeltaOps) {
+    const diff = getDiffEntry(data);
+
+    const res = await this.db.collection(this.collectionName).insertOne({
+      op: op.toString(),
+      ...diff
+    })
+
+    return res.insertedId;
+  }
+
+  public async sync(lastSync: number, filter?: FilterQuery<any>) {
+    const syncPipeline = [
+      {
+        $match: {
+          timestamp: {
+            $gte: new Date(lastSync)
+          }
+
+        },
+      },
+      {
+        $group: {
+          _id: "$docId",
+          sets: {
+            $mergeObjects: "$sets"
+          },
+          lastOp: {
+            $last: "$op"
+          },
+          lastUpdated: {
+            $last: "$timestamp"
+          }
+        }
+      }, {
+        $replaceRoot: {
+          newRoot: {
+            $mergeObjects: [
+              {
+                _id: "$_id",
+                _deleted: { $eq: ["$lastOp", "delete"] },
+                updatedAt: {
+                  $toLong: "$lastUpdated"
+                }
+              },
+              "$sets"
+            ]
+          }
+        }
+      }, {
+        $match: filter
+      }];
+
+    return this.db.collection(this.collectionName).aggregate(syncPipeline).toArray()
+  }
+
+  public async checkForConflicts(updateData: any): Promise<boolean> {
+    if (updateData === {} || updateData === undefined) {
```

Comment:
Because of object references, I am not sure if `if (updateData === {})` would work.

A search of "how do check if object is empty in javascript" shows no examples using this method. Best to look into it a bit more

https://stackoverflow.com/questions/4994201/is-object-empty